Implement Operating Levers, Decorative Bronze gate in Halvung. Correc…

hooksta4 opened the pull request:
…t Bracelet of Verve  to correct QM. Addresses Issue #754 

Adjusted name for Operating Levers in npc_list.sql
Implemented correct QM for Bracelet of Verve
Corrected script for obtaining KI Bracelet of Verve (deleted incorrect script QM5)
Added messages to IDs.lua
Added items to Items.lua

<!-- remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm: -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

hooksta4 committed:
Implement Operating Levers, Decorative Bronze gate in Halvung. Correct Bracelet of Verve  to correct QM.

Adjusted name for Operating Levers in npc_list.sql
Implemented correct QM for Bracelet of Verve
Corrected script for obtaining KI Bracelet of Verve (deleted incorrect script QM5)
Added messages to IDs.lua
Added items to Items.lua

hooksta4 committed:
Update to Troll Cameist, Troll Engraver, Troll Gemologist mob_droplist tables

Updated Troll Cameist, Troll Engraver, Troll Gemologist drop tables to reflect
ffxidb drop rates (avg) implemented drops for issue #754
This should close out issue, pending any corrections?

hooksta4 committed:
Trailing white spaces in IDs.lua

facepalm

TeoTwawki commented on scripts/zones/Halvung/npcs/Operating_Lever_A.lua:
```diff
@@ -0,0 +1,30 @@
+end
+
+entity.onTrigger = function(player, npc)
+    if player:hasKeyItem(xi.ki.BRACELET_OF_VERVE) == false then
```
`== false` is not needed

hooksta4 replied:
I changed the logic around to reflect not using ==false.

TeoTwawki commented on scripts/zones/Halvung/npcs/Operating_Lever_A.lua:
```diff
@@ -0,0 +1,30 @@
+    else
+        GetNPCByID(npc:getID() - 2):openDoor(30)
+        player:messageSpecial(ID.text.LIFT_LEVER)
+        return 1
```
Don't use returns like this in onTrigger. There is no need for them.

TeoTwawki commented on scripts/zones/Halvung/npcs/_1qp.lua:
```diff
@@ -0,0 +1,37 @@
+end
+
+entity.onTrigger = function(player, npc)
+    if player:getZPos() <= 79 and npc:getAnimation() == xi.anim.CLOSE_DOOR then --from inside the door
```
comments are started by `-- ` not `--` (note the space) oir legibility reasons in lua, and for execution failure reasons in sql.

TeoTwawki replied:
Also going to want this really well tested before settling on just a Z coordinate check. Not sure what best practice here should be but feels like we should have something more.

hooksta4 replied:
I too, was curious about this as well on how to check if inside or outside, a lot of other doors that came to mind were banishing gate and Den of Rancor switch. Both of which have Z checks.

TeoTwawki replied:
yeah figuring this out is a work in progress for us, we've been trying to reduce that kind of coordinate checking. Sometimes its unavoidable.

hooksta4 replied:
how should i proceed with this then? KO offered some advice below: will try that out here in a little bit and give some feedback.

TeoTwawki replied:
be very thorough on testing both sides of that game, maybe also include other axis's. and give some time for others to chime in with any ideas.

hooksta4 replied:
ok, i think from what I have done so far the Z will suffice because how the area is setup... I will record and post from opposite side and inside; will provide link later.

hooksta4 replied:
I think what will make this specifically work is how thick the door is but as I said more will be done. Thanks

hooksta4 replied:
So after a lot of testing and !pos myself I feel confident that if player:getZPos() <= 79.75 is a solid place, which you can only get to by doing !pos. From the locked side you cannot get any closer than 80.8981

TeoTwawki commented on scripts/zones/Halvung/npcs/_1qp.lua:
```diff
@@ -0,0 +1,37 @@
+local entity = {}
+
+entity.onTrade = function(player, npc, trade)
+    if npcUtil.tradeHas(trade, xi.items.HALVUNG_SHAKUDO_KEY) and
```
```lua
    if
        npcUtil.tradeHas(trade, xi.items.HALVUNG_SHAKUDO_KEY) and
        npcUtil.tradeHas(trade, xi.items.HALVUNG_BRONZE_KEY) and
        npcUtil.tradeHas(trade, xi.items.HALVUNG_BRASS_KEY)
    then
```
per the style guide

KnowOne134 replied:
npcUtil.tradeHasExactly(trade, {item1, item2, item3})

TeoTwawki replied:
what he said and then you don't have to break a wide line in the 1st place. I'm still not used to these utils _existing_

..old grandpa Teo had to explicitly ask whats was in the trade boxes back in his day.

hooksta4 replied:
> npcUtil.tradeHasExactly(trade, {item1, item2, item3})

it seems this removes the ability to use the items.lua?
    if npcUtil.tradeHasExactly(trade, {2221, 2222, 2223}) then
vice     if npcUtil.tradeHasExactly(trade, {HALVUNG_SHAKUDO_KEY, HALVUNG_BRONZE_KEY, HALVUNG_BRASS_KEY}) then

TeoTwawki replied:
`if npcUtil.tradeHasExactly(trade, {xi.items.HALVUNG_SHAKUDO_KEY, xi.items.HALVUNG_BRONZE_KEY, xi.items.HALVUNG_BRASS_KEY}) then`

Also I now see its still a bit wide.

hooksta4 replied:
It wouldn't trade with the latter example, pushed with 2221,2222,2223 as that did work when testing it.

TeoTwawki replied:
you were missing `xi.items.` in your post though. I -know- it works, so check again you had to make a mistake

TeoTwawki replied:
the human readable form also works with the `messageSpecial()` call below it

hooksta4 replied:
That makes sense.

TeoTwawki commented on scripts/zones/Halvung/npcs/qm8.lua:
```diff
@@ -1,5 +1,5 @@
 -----------------------------------
--- Area: Halvung
+--  Area: Halvung
```
no need to adjust the whitespace here

TeoTwawki commented on scripts/zones/Halvung/npcs/Operating_Lever_B.lua:
```diff
@@ -0,0 +1,30 @@
+entity.onTrigger = function(player, npc)
+    if player:hasKeyItem(xi.ki.BRACELET_OF_VERVE) == false then
+        player:startEvent(100)
+        return 1
```
Same things I mentioned in `Operating Lever A` here also

TeoTwawki commented on scripts/zones/Halvung/npcs/_1qp.lua:
```diff
@@ -0,0 +1,37 @@
+require("scripts/globals/npc_util")
+require("scripts/globals/items")
+require("scripts/globals/status")
+
```
line break on wrong side of the dashed line.

hooksta4 commented:
Thanks for the feedback! Will make adjustments to maintain style guide and resubmit. Again, thanks for all y'all do.

hooksta4 committed:
Style and logic changes as per the requested changes.

This time with less spaces

KnowOne134 commented on scripts/zones/Halvung/npcs/_1qp.lua:
```diff
@@ -0,0 +1,34 @@
+end
+
+entity.onTrigger = function(player, npc)
+    if player:getZPos() <= 79 and npc:getAnimation() == xi.anim.CLOSE_DOOR then -- from inside the door
```
Thought not sure if it works like does the invisible door turn to face you are not but if not
If npc:isFacing(player) then or
If not npc:isFacing(player) then

If this does work could fix all them doorways to Staging points

hooksta4 replied:
which invisible door(s) are you referring to?

KnowOne134 replied:
Don't mind me I was thinking out loud

TeoTwawki commented on scripts/zones/Halvung/npcs/_1qp.lua:
```diff
@@ -0,0 +1,34 @@
+    if npcUtil.tradeHasExactly(trade, {2221, 2222, 2223}) then
+        player:confirmTrade()
+        npc:openDoor()
+        player:messageSpecial(ID.text.KEY_BREAKS, 2221,2222,2223)
```
still need these magic numbers out.
```lua
entity.onTrade = function(player, npc, trade)
    if
        npcUtil.tradeHasExactly(trade,
        {
            xi.items.HALVUNG_SHAKUDO_KEY,
            xi.items.HALVUNG_BRONZE_KEY,
            xi.items.HALVUNG_BRASS_KEY
        })
    then
        player:confirmTrade()
        npc:openDoor()
        player:messageSpecial(ID.text.KEY_BREAKS, 
            xi.items.HALVUNG_SHAKUDO_KEY,
            xi.items.HALVUNG_BRONZE_KEY,
            xi.items.HALVUNG_BRASS_KEY
        )
    end
end
```
Unsure of proper formatting of these widelines, this is my best guess.

@claywar @zach2good this doesn't exactly fit with examples in our style guide, I think. Care to chime in?

hooksta4 replied:
Yes, I am working literally on that right now ;) Testing

hooksta4 committed:
Updated to use xi.items for keys, and keybreak message, adjusted Zpos on inside door.

claywar commented on scripts/zones/Halvung/npcs/Operating_Lever_A.lua:
```diff
@@ -0,0 +1,28 @@
+
+entity.onTrigger = function(player, npc)
+    if player:hasKeyItem(xi.ki.BRACELET_OF_VERVE) then
+        GetNPCByID(npc:getID() - 2):openDoor(30)
```
This seems potentially brittle.  I think I'd feel better defining the ID in IDs.lua under the npc section, that way if anything is inserted between in the future we're safe.

hooksta4 replied:
Would you kindly elaborate a little bit on brittle? 

MINING = 
{
}

KEYITEM = 
{
BRACELET_OF_VERVE = 2579
}

assuming that is for if ID's shift at some point? 


Edit: disregard was tired last night ha. I get what you mean now.

hooksta4 replied:
@claywar So thinking on this a little bit, this specific scenario has 10 levers, which when triggered open up the door (which is its own npcid). Would it be more efficient, long term, to name the doors by ID?
Example: 1st set of doors
Operating_Lever_A = 17031670 
Operating_Lever_B = 17031671
when triggered both those doors will open 17031668 (_1qf) 

Second set of doors
Operating_Lever_A = 17031674 
Operating_Lever_B = 1703167
when triggered both those doors will open 17031672 (_1qf) 

etc..

With that said, I could rename all the operating levers to Operating_Lever_A-J in npc_list.sql, define each actual door in IDs.lua (DOOR_A_B),  = 17031668, DOOR_C_D = 17031672 etc., and add lua for each of the Operating_Lever_A-J, to open the corresponding door?

`INSERT INTO `npc_list` VALUES (17031668,'_1qf','_1qf',

INSERT INTO `npc_list` VALUES (17031669,'_1qg','

INSERT INTO `npc_list` VALUES (17031670,'Operating_Lever_A','Operating Lever'

INSERT INTO `npc_list` VALUES (17031671,'Operating_Lever_B','Operating Lever',

`

I believe as long as they shift together it should be fine yea?

hooksta4 replied:
> This seems potentially brittle.  I think I'd feel better defining the ID in IDs.lua under the npc section, that way if anything is inserted between in the future we're safe.

If you mean by a gap (inserted) in between the ids, there isn't one as there is another npcid that is actually in the door frame but does nothing. ?? Thoughts

Xaver-DaRed replied:
npcs get shifted every month lately. It isn't a matter of all shifting +-X. Npcs get thrown in the middle for no reason.

With that in mind, I agree the NPCs should be defined in zones/Halvung/IDs file and just call them like you call text IDs

claywar commented on scripts/zones/Halvung/npcs/Operating_Lever_B.lua:
```diff
@@ -0,0 +1,28 @@
+
+entity.onTrigger = function(player, npc)
+    if player:hasKeyItem(xi.ki.BRACELET_OF_VERVE) then
+        GetNPCByID(npc:getID() - 3):openDoor(30)
```
Same as above

claywar commented on scripts/zones/Halvung/npcs/_1qp.lua:
```diff
@@ -0,0 +1,45 @@
+entity.onTrade = function(player, npc, trade)
+    if
+        npcUtil.tradeHasExactly(trade,
+        {
```
Styling-wise, I can be on board with this, but I'd probably prefer breaking out the block on the `(` level

claywar replied:
Disregard this comment

hooksta4 committed:
Merge branch 'LandSandBoat:base' into halvung_doors

Xaver-DaRed commented on sql/mob_droplist.sql:
```diff
@@ -12883,16 +12883,22 @@ INSERT INTO `mob_droplist` VALUES (2472,0,0,1000,3225,30);
+INSERT INTO `mob_droplist` VALUES (2474,0,0,1000,2160,28);
+INSERT INTO `mob_droplist` VALUES (2474,0,0,1000,2223,74);
+INSERT INTO `mob_droplist` VALUES (2474,0,0,1000,2333,213);
+INSERT INTO `mob_droplist` VALUES (2474,0,0,1000,8801,73);
```
This comment applies to all mob drops. They should conform to the chart I linked in your other PR:
https://github.com/LandSandBoat/server/pull/1129

hooksta4 replied:
Got it. For those that do not have a label however, is there a standard that should be used then?

TeoTwawki replied:
> so compared with other drops from other mobs in same zone to reflect the same 

no, bad newbie :newspaper_roll: stop that and wait till ppl have time to answer the question

hooksta4 committed:
Updated doors to id.npc, droplist to reflect SE table, added lua's for doors.

Updated lot tables to reflect SE drop tables, some were not listed as U,V,R etc, so compared with other drops from other mobs in same zone to reflect the same %. Renamed each door as A-J, added ID.npc.Lever_AB-IJ into npc.lua as well as associated LUA's to open correct door. And created Lever_AB_Door - Lever_IJ_Door in IDs.lua to handle future id shifts.
